packages/runed/src/lib/utilities/PersistedState/PersistedState.svelte.ts

```diff
@@ -0,0 +1,203 @@
+		const listener = (event: StorageEvent) => {
+			if (event.key === null) {
+				return;
+			}
+
+			const result: GetItemResult<T> =
+				event.newValue !== null
+					? { found: true, value: this.#serializer.deserialize(event.newValue) }
+					: { found: false, value: null };
+
+			callback(event.key, result);
+		};
+
+		const unsubscribe = addEventListener(window, "storage", listener.bind(this));
+
+		return () => {
+			unsubscribe();
+		};
+	}
+}
+
+async function setValueToStorage<T>({
+	key,
+	value,
+	storage,
+}: {
+	key: string;
+	value: T;
+	storage: StorageAdapter<T> | null;
+}) {
+	if (!storage) {
+		return;
+	}
+
+	try {
+		await storage.setItem(key, value);
+	} catch (e) {
+		console.error(
+			`Error when writing value from persisted store "${key}" to ${storage.constructor.name}`,
+			e
+		);
+	}
+}
+
+function getWebStorageAdapterForStorageType<T>(storageType: StorageType): StorageAdapter<T> | null {
+	if (typeof window === "undefined") {
+		return null;
+	}
+
+	const webStorageAdapterByStorageType = {
+		local: new WebStorageAdapter<T>({ storage: localStorage }),
+		session: new WebStorageAdapter<T>({ storage: sessionStorage }),
+	};
+
+	return webStorageAdapterByStorageType[storageType];
+}
+
+type StorageType = "local" | "session";
+
+type PersistedStateOptions<T> = {
+	/** The storage type to use. Defaults to `local`. */
+	storage?: StorageType | StorageAdapter<T>;
+	/**
+	 * Whether to synchronize the state with changes detected via the `subscribe` callback. This allows the state to be updated in response to changes originating from various sources, including other browser tabs or sessions when using web storage like localStorage or sessionStorage. Defaults to `true`.
+	 */
+	syncChanges?: boolean;
+};
+
+/**
+ * Creates reactive state that is persisted and synchronized across browser sessions and tabs using Web Storage.
+ * @param key The unique key used to store the state in the storage.
+ * @param initialValue The initial value of the state if not already present in the storage.
+ * @param options Configuration options including storage type, serializer for complex data types, and whether to sync state changes across tabs.
+ *
+ * @see {@link https://runed.dev/docs/utilities/persisted-state}
+ */
+export class Persisted<T> {
+	#current = $state() as T;
+	#isInitialized = $state(false);
+	#initialValue: T;
+	#key: string;
+	#storageAdapter: StorageAdapter<T> | null;
+
+	constructor(key: string, initialValue: T, options: PersistedStateOptions<T> = {}) {
+		const { storage = "local", syncChanges = true } = options;
+
+		this.#key = key;
+		this.#initialValue = initialValue;
+		this.#storageAdapter =
+			typeof storage === "string" ? getWebStorageAdapterForStorageType(storage) : storage;
+
+		$effect(() => {
+			if (!this.#isInitialized) {
+				return;
+			}
+
+			setValueToStorage({
+				key: this.#key,
+				value: this.#current,
+				storage: this.#storageAdapter,
+			});
+		});
+
+		if (syncChanges) {
```

Comment:
I haven't stumbled upon such a scenario in a real application.  However, sounds interesting.  Maybe is good for those creative thinkers out there that may use local storage to sync multiple windows.  This could be used from debugging to fancier connect/disconnect windows.